src/contracts/rewarder/ODefaultOperatorRewards.sol

```diff
@@ -0,0 +1,235 @@
+            uint256 balanceBefore = IERC20(i_token).balanceOf(address(this));
+            IERC20(i_token).safeTransferFrom(msg.sender, address(this), amount);
+            amount = IERC20(i_token).balanceOf(address(this)) - balanceBefore;
+
+            if (amount == 0) {
+                revert ODefaultOperatorRewards__InsufficientTransfer();
+            }
+
+            if (totalPointsToken == 0) {
+                revert ODefaultOperatorRewards__InvalidTotalPoints();
+            }
+        }
+
+        uint256 tokensPerPoint = amount / totalPointsToken; // TODO: To change/check the math.
+
+        EraRoot memory eraRoot = EraRoot({epoch: epoch, amount: amount, tokensPerPoint: tokensPerPoint, root: root});
+
+        s_eraRoot[eraIndex] = eraRoot;
+        s_eraIndexesPerEpoch[epoch].push(eraIndex);
+
+        emit DistributeRewards(eraIndex, epoch, tokensPerPoint, amount, root);
+    }
+
+    /**
+     * @inheritdoc IODefaultOperatorRewards
+     */
+    function claimRewards(
+        ClaimRewardsInput calldata input
+    ) external nonReentrant returns (uint256 amount) {
+        bytes32 root_ = s_eraRoot[input.eraIndex].root;
+        if (root_ == bytes32(0)) {
+            revert ODefaultOperatorRewards__RootNotSet();
+        }
+
+        if (
+            !MerkleProof.verifyCalldata(
+                input.proof,
+                root_,
+                keccak256(abi.encodePacked(input.operatorKey, ScaleCodec.encodeU32(input.totalPointsClaimable)))
+            )
+        ) {
+            revert ODefaultOperatorRewards__InvalidProof();
+        }
+
+        address middlewareAddress = INetworkMiddlewareService(i_networkMiddlewareService).middleware(i_network);
+        address recipient = SimpleKeyRegistry32(middlewareAddress).getOperatorByKey(input.operatorKey);
+
+        // Calculate the total amount of tokens that can be claimed which is:
+        // total amount of tokens = total points claimable * tokens per point
+        amount = input.totalPointsClaimable * s_eraRoot[input.eraIndex].tokensPerPoint;
+
+        // You can only claim everything and if it's claimed before revert
+        if (s_claimed[input.eraIndex][recipient] > 0) {
+            revert ODefaultOperatorRewards__InsufficientTotalClaimable();
+        }
+
+        s_claimed[input.eraIndex][recipient] = amount;
+
+        // s_operatorShare% of the rewards to the operator
+        uint256 operatorAmount = amount.mulDiv(s_operatorShare, 100);
+
+        // (1-s_operatorShare)% of the rewards to the stakers
+        uint256 stakerAmount = amount - operatorAmount;
+
+        // On every claim send s_operatorShare% of the rewards to the operator
+        // And then distribute rewards to the stakers
+        // This is gonna send (1-s_operatorShare)% of the rewards
+        IERC20(i_token).safeTransfer(recipient, operatorAmount);
+
+        _distributeRewardsToStakers(input.epoch, input.eraIndex, stakerAmount, recipient, middlewareAddress, input.data);
```

Comment:
epoch should not be in the input, but rather, we should fetch it from storage. Otherwise it can be gamed